// @solana/spl-token-registry/dist/module/lib/tokens.js
import devnet from './../tokens/devnet.json';
import mainnetBeta from './../tokens/mainnet-beta.json';
import testnet from './../tokens/testnet.json';
import * as cross from 'cross-fetch';
export var ENV;
(function (ENV) {
    ENV["MainnetBeta"] = "mainnet-beta";
    ENV["Testnet"] = "testnet";
    ENV["Devnet"] = "devnet";
})(ENV || (ENV = {}));
export const TOKENS = {
    [ENV.MainnetBeta]: mainnetBeta,
    [ENV.Testnet]: testnet,
    [ENV.Devnet]: devnet,
};
export class SolanaTokenListResolutionStrategy {
    constructor() {
        this.resolve = async (network) => {
            throw new Error(`Not Implemented Yet. ${network}`);
        };
    }
}
export class StaticTokenListResolutionStrategy {
    constructor() {
        this.resolve = async (network) => {
            return TOKENS[network];
        };
    }
}
const queryJsonFiles = async (network, files) => {
    const responses = await Promise.all(files.map(async (repo) => {
        try {
            const response = await cross.fetch(`${repo}/${network}.json`);
            const json = (await response.json());
            return json;
        }
        catch {
            return [];
        }
    }));
    return responses.reduce((acc, arr) => acc.concat(arr), []);
};
export class GitHubTokenListResolutionStrategy {
    constructor() {
        this.repositories = [
            'https://raw.githubusercontent.com/solana-labs/token-list/main/src/tokens',
            'https://raw.githubusercontent.com/project-serum/serum-ts/master/packages/tokens/src',
        ];
        this.resolve = async (network) => {
            return queryJsonFiles(network, this.repositories);
        };
    }
}
export class CDNTokenListResolutionStrategy {
    constructor() {
        this.repositories = [
            'https://cdn.jsdelivr.net/gh/solana-labs/token-list@main/src/tokens',
            'https://cdn.jsdelivr.net/gh/project-serum/serum-ts@master/packages/tokens/src',
        ];
        this.resolve = async (network) => {
            return queryJsonFiles(network, this.repositories);
        };
    }
}
export var Strategy;
(function (Strategy) {
    Strategy["GitHub"] = "GitHub";
    Strategy["Static"] = "Static";
    Strategy["Solana"] = "Solana";
    Strategy["CDN"] = "CDN";
})(Strategy || (Strategy = {}));
export class TokenListProvider {
    constructor() {
        this.resolve = async (network, strategy = Strategy.CDN) => {
            return TokenListProvider.strategies[strategy].resolve(network);
        };
    }
}
TokenListProvider.strategies = {
    [Strategy.GitHub]: new GitHubTokenListResolutionStrategy(),
    [Strategy.Static]: new StaticTokenListResolutionStrategy(),
    [Strategy.Solana]: new SolanaTokenListResolutionStrategy(),
    [Strategy.CDN]: new CDNTokenListResolutionStrategy(),
};
